In [81]:
from siuba import _,select,filter
import pandas as pd
import numpy as np


In [82]:
# using pandas to read the excel file.
df = pd.read_excel(r"C:\Users\user\OneDrive\Documents\UOPEOPLE COURSES\DATA SCIENCE\Excel data\data\itunes_data.xls")

# using siuba to check the data
df >> _.head()

a                                   Track  \
0  0  All the Best Cowboys Have Daddy Issues   
1  1                               Beira Mar   
2  2                                  Brasil   
3  3                            Ben Franklin   
4  4            O Último Romântico (Ao Vivo)   

                            Composer  Milliseconds      Bytes  UnitPrice  \
0                                NaN       2555492  211743651       1.99   
1                       Gilberto Gil        295444    9597994       0.99   
2  Milton Nascimento, Fernando Brant        155428    5252560       0.99   
3                                NaN       1271938  264168080       1.99   
4                                NaN        231993    7692697       0.99   

      Genre                                            Album  \
0  TV Shows                                   Lost, Season 1   
1     Latin                                        Unplugged   
2     Latin                        Milton Nascimento Ao Vivo   
3    Comedy                             The Office, Season 3   
4     Latin  Lulu Santos - RCA 100 Anos De Música - Álbum 02   

              Artist  
0               Lost  
1       Eric Clapton  
2  Milton Nascimento  
3         The Office  
4        Lulu Santos

## **Using Siuba and Pandas to analyze itunes data**

Itunes is a popular music steaming service used around the world. The code takes data from an excel file containing the popular tracks, composer, time spent listening to a particular song and the genre as well. 

The goal if this analysis is to process and clean the data, preparing the data for *visulaization*. Each cell explains the code and the processes being done.

In [83]:
#filtering out second index

df = df >> select(~_.a)

df >> _.head()

# using siuba select to check each column for missing data
df.info()
df.shape
df.count()

df >> _.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Track         303 non-null    object 
 1   Composer      227 non-null    object 
 2   Milliseconds  303 non-null    int64  
 3   Bytes         303 non-null    int64  
 4   UnitPrice     303 non-null    float64
 5   Genre         303 non-null    object 
 6   Album         303 non-null    object 
 7   Artist        303 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 19.1+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Track         303 non-null    object 
 1   Composer      227 non-null    object 
 2   Milliseconds  303 non-null    int64  
 3   Bytes         303 non-null    int64  
 4   UnitPrice     303 non-null    float6

###### We can see that data above contains some missing values in the composer's column, hence it will be dealt with.

In [84]:
# filling unknown values in Composer columns
df["Composer"] = df >> select(_.Composer)>> _.fillna("Unknown")

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Track         303 non-null    object 
 1   Composer      303 non-null    object 
 2   Milliseconds  303 non-null    int64  
 3   Bytes         303 non-null    int64  
 4   UnitPrice     303 non-null    float64
 5   Genre         303 non-null    object 
 6   Album         303 non-null    object 
 7   Artist        303 non-null    object 
dtypes: float64(1), int64(2), object(5)
memory usage: 19.1+ KB


In [85]:
# Change column name to songtrack and Make track column the index

from siuba import rename

df = df >> rename(Song_Track = _.Track) >> _.head()


# df['Song_Track'] = df >> select(_.Song_Track == _.Track)

# df = df >> select(~ _.Track)

df = df.set_index('Song_Track')

df >> _.head()


Composer  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues                            Unknown   
Beira Mar                                                    Gilberto Gil   
Brasil                                  Milton Nascimento, Fernando Brant   
Ben Franklin                                                      Unknown   
O Último Romântico (Ao Vivo)                                      Unknown   

                                        Milliseconds      Bytes  UnitPrice  \
Song_Track                                                                   
All the Best Cowboys Have Daddy Issues       2555492  211743651       1.99   
Beira Mar                                     295444    9597994       0.99   
Brasil                                        155428    5252560       0.99   
Ben Franklin                                 1271938  264168080       1.99   
O Último Romântico (Ao Vivo)                  231993    7692697       0.99   

                                           Genre  \
Song_Track                                         
All the Best Cowboys Have Daddy Issues  TV Shows   
Beira Mar                                  Latin   
Brasil                                     Latin   
Ben Franklin                              Comedy   
O Último Romântico (Ao Vivo)               Latin   

                                                                                  Album  \
Song_Track                                                                                
All the Best Cowboys Have Daddy Issues                                   Lost, Season 1   
Beira Mar                                                                     Unplugged   
Brasil                                                        Milton Nascimento Ao Vivo   
Ben Franklin                                                       The Office, Season 3   
O Último Romântico (Ao Vivo)            Lulu Santos - RCA 100 Anos De Música - Álbum 02   

                                                   Artist  
Song_Track                                                 
All the Best Cowboys Have Daddy Issues               Lost  
Beira Mar                                    Eric Clapton  
Brasil                                  Milton Nascimento  
Ben Franklin                                   The Office  
O Último Romântico (Ao Vivo)                  Lulu Santos

In [86]:
# Changing unit price name to indicate currency as well
df["Unit_Price_dollars"] = df['UnitPrice']

df.head()

df = df >> select(~_.UnitPrice)

df.head()

Composer  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues                            Unknown   
Beira Mar                                                    Gilberto Gil   
Brasil                                  Milton Nascimento, Fernando Brant   
Ben Franklin                                                      Unknown   
O Último Romântico (Ao Vivo)                                      Unknown   

                                        Milliseconds      Bytes     Genre  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues       2555492  211743651  TV Shows   
Beira Mar                                     295444    9597994     Latin   
Brasil                                        155428    5252560     Latin   
Ben Franklin                                 1271938  264168080    Comedy   
O Último Romântico (Ao Vivo)                  231993    7692697     Latin   

                                                                                  Album  \
Song_Track                                                                                
All the Best Cowboys Have Daddy Issues                                   Lost, Season 1   
Beira Mar                                                                     Unplugged   
Brasil                                                        Milton Nascimento Ao Vivo   
Ben Franklin                                                       The Office, Season 3   
O Último Romântico (Ao Vivo)            Lulu Santos - RCA 100 Anos De Música - Álbum 02   

                                                   Artist  Unit_Price_dollars  
Song_Track                                                                     
All the Best Cowboys Have Daddy Issues               Lost                1.99  
Beira Mar                                    Eric Clapton                0.99  
Brasil                                  Milton Nascimento                0.99  
Ben Franklin                                   The Office                1.99  
O Último Romântico (Ao Vivo)                  Lulu Santos                0.99

In [87]:
# Creating minutes column from miiliseonds to minutes to make it comprehensible

df['Total_runtime_min'] = df >> select(_.Milliseconds)

df.head()

# 60000 milliseconds = 1 minute 
df['Total_runtime_min'] = np.divide(df['Total_runtime_min'],60000)

df.head()

Composer  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues                            Unknown   
Beira Mar                                                    Gilberto Gil   
Brasil                                  Milton Nascimento, Fernando Brant   
Ben Franklin                                                      Unknown   
O Último Romântico (Ao Vivo)                                      Unknown   

                                        Milliseconds      Bytes     Genre  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues       2555492  211743651  TV Shows   
Beira Mar                                     295444    9597994     Latin   
Brasil                                        155428    5252560     Latin   
Ben Franklin                                 1271938  264168080    Comedy   
O Último Romântico (Ao Vivo)                  231993    7692697     Latin   

                                                                                  Album  \
Song_Track                                                                                
All the Best Cowboys Have Daddy Issues                                   Lost, Season 1   
Beira Mar                                                                     Unplugged   
Brasil                                                        Milton Nascimento Ao Vivo   
Ben Franklin                                                       The Office, Season 3   
O Último Romântico (Ao Vivo)            Lulu Santos - RCA 100 Anos De Música - Álbum 02   

                                                   Artist  Unit_Price_dollars  \
Song_Track                                                                      
All the Best Cowboys Have Daddy Issues               Lost                1.99   
Beira Mar                                    Eric Clapton                0.99   
Brasil                                  Milton Nascimento                0.99   
Ben Franklin                                   The Office                1.99   
O Último Romântico (Ao Vivo)                  Lulu Santos                0.99   

                                        Total_runtime_min  
Song_Track                                                 
All the Best Cowboys Have Daddy Issues          42.591533  
Beira Mar                                        4.924067  
Brasil                                           2.590467  
Ben Franklin                                    21.198967  
O Último Romântico (Ao Vivo)                     3.866550

In [88]:
# Sort out playtime from most played to least played

from siuba import arrange

# using arrange verb to sort playtime in descending order
df = df >> arrange(-_.Total_runtime_min)

df.head()

Composer  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues                            Unknown   
Ben Franklin                                                      Unknown   
Beira Mar                                                    Gilberto Gil   
O Último Romântico (Ao Vivo)                                      Unknown   
Brasil                                  Milton Nascimento, Fernando Brant   

                                        Milliseconds      Bytes     Genre  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues       2555492  211743651  TV Shows   
Ben Franklin                                 1271938  264168080    Comedy   
Beira Mar                                     295444    9597994     Latin   
O Último Romântico (Ao Vivo)                  231993    7692697     Latin   
Brasil                                        155428    5252560     Latin   

                                                                                  Album  \
Song_Track                                                                                
All the Best Cowboys Have Daddy Issues                                   Lost, Season 1   
Ben Franklin                                                       The Office, Season 3   
Beira Mar                                                                     Unplugged   
O Último Romântico (Ao Vivo)            Lulu Santos - RCA 100 Anos De Música - Álbum 02   
Brasil                                                        Milton Nascimento Ao Vivo   

                                                   Artist  Unit_Price_dollars  \
Song_Track                                                                      
All the Best Cowboys Have Daddy Issues               Lost                1.99   
Ben Franklin                                   The Office                1.99   
Beira Mar                                    Eric Clapton                0.99   
O Último Romântico (Ao Vivo)                  Lulu Santos                0.99   
Brasil                                  Milton Nascimento                0.99   

                                        Total_runtime_min  
Song_Track                                                 
All the Best Cowboys Have Daddy Issues          42.591533  
Ben Franklin                                    21.198967  
Beira Mar                                        4.924067  
O Último Romântico (Ao Vivo)                     3.866550  
Brasil                                           2.590467

In [89]:
# Changing bytes into megabytes for easier comprehension

df['song_size_Mb'] = df["Bytes"]

df.head()

# Converting bytes to megabytes. 1 Mb = (1024)2 bytes
bytes_megabytes = np.multiply (1024,1024)

df['song_size_Mb'] = np.divide(df['song_size_Mb'], bytes_megabytes)

df.head()

Composer  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues                            Unknown   
Ben Franklin                                                      Unknown   
Beira Mar                                                    Gilberto Gil   
O Último Romântico (Ao Vivo)                                      Unknown   
Brasil                                  Milton Nascimento, Fernando Brant   

                                        Milliseconds      Bytes     Genre  \
Song_Track                                                                  
All the Best Cowboys Have Daddy Issues       2555492  211743651  TV Shows   
Ben Franklin                                 1271938  264168080    Comedy   
Beira Mar                                     295444    9597994     Latin   
O Último Romântico (Ao Vivo)                  231993    7692697     Latin   
Brasil                                        155428    5252560     Latin   

                                                                                  Album  \
Song_Track                                                                                
All the Best Cowboys Have Daddy Issues                                   Lost, Season 1   
Ben Franklin                                                       The Office, Season 3   
Beira Mar                                                                     Unplugged   
O Último Romântico (Ao Vivo)            Lulu Santos - RCA 100 Anos De Música - Álbum 02   
Brasil                                                        Milton Nascimento Ao Vivo   

                                                   Artist  Unit_Price_dollars  \
Song_Track                                                                      
All the Best Cowboys Have Daddy Issues               Lost                1.99   
Ben Franklin                                   The Office                1.99   
Beira Mar                                    Eric Clapton                0.99   
O Último Romântico (Ao Vivo)                  Lulu Santos                0.99   
Brasil                                  Milton Nascimento                0.99   

                                        Total_runtime_min  song_size_Mb  
Song_Track                                                               
All the Best Cowboys Have Daddy Issues          42.591533    201.934482  
Ben Franklin                                    21.198967    251.930313  
Beira Mar                                        4.924067      9.153360  
O Último Romântico (Ao Vivo)                     3.866550      7.336328  
Brasil                                           2.590467      5.009232

In [90]:
# Checking some statistics using numpy

# The mean price of the songs available
np.mean(df['Unit_Price_dollars'])



1.3900000000000001

###### About price mean

The average price of a song on this list is $1.39. 

In [91]:
#The mean for size of the songs
np.mean(df['song_size_Mb'])

95.07274284362794

In [92]:
df >> _.describe()

Milliseconds         Bytes  Unit_Price_dollars  Total_runtime_min  \
count  5.000000e+00  5.000000e+00            5.000000           5.000000   
mean   9.020590e+05  9.969100e+07            1.390000          15.034317   
std    1.030180e+06  1.275809e+08            0.547723          17.169658   
min    1.554280e+05  5.252560e+06            0.990000           2.590467   
25%    2.319930e+05  7.692697e+06            0.990000           3.866550   
50%    2.954440e+05  9.597994e+06            0.990000           4.924067   
75%    1.271938e+06  2.117437e+08            1.990000          21.198967   
max    2.555492e+06  2.641681e+08            1.990000          42.591533   

       song_size_Mb  
count      5.000000  
mean      95.072743  
std      121.670645  
min        5.009232  
25%        7.336328  
50%        9.153360  
75%      201.934482  
max      251.930313

###### About Song size and runtime

The average size of on this list song is 22 megabytes. It seems there are some outliers increasing the song size. 

Upon inspection, it seems there are some episodes on the list meaning that movies,TV episodes and other programs were included in this list.
We used describe() to find the minimum and maximum values and compare to confirm this theory 


We may run into this problem with the total mean runtime as well.

To solve this problem, we could filter out any song track having a runtime greater than 10 minutes, to reflect the 75% percentile's number of 5 minutes. Care must be taken not to casually filter any important data.
The *filter()* verb in ***siuba*** could be used for this

In [93]:
df = df >> filter(_.Total_runtime_min <= 10)

df.head()

Composer  Milliseconds  \
Song_Track                                                                      
Beira Mar                                          Gilberto Gil        295444   
O Último Romântico (Ao Vivo)                            Unknown        231993   
Brasil                        Milton Nascimento, Fernando Brant        155428   

                                Bytes  Genre  \
Song_Track                                     
Beira Mar                     9597994  Latin   
O Último Romântico (Ao Vivo)  7692697  Latin   
Brasil                        5252560  Latin   

                                                                        Album  \
Song_Track                                                                      
Beira Mar                                                           Unplugged   
O Último Romântico (Ao Vivo)  Lulu Santos - RCA 100 Anos De Música - Álbum 02   
Brasil                                              Milton Nascimento Ao Vivo   

                                         Artist  Unit_Price_dollars  \
Song_Track                                                            
Beira Mar                          Eric Clapton                0.99   
O Último Romântico (Ao Vivo)        Lulu Santos                0.99   
Brasil                        Milton Nascimento                0.99   

                              Total_runtime_min  song_size_Mb  
Song_Track                                                     
Beira Mar                              4.924067      9.153360  
O Último Romântico (Ao Vivo)           3.866550      7.336328  
Brasil                                 2.590467      5.009232

In [94]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3 entries, Beira Mar to Brasil
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Composer            3 non-null      object 
 1   Milliseconds        3 non-null      int64  
 2   Bytes               3 non-null      int64  
 3   Genre               3 non-null      object 
 4   Album               3 non-null      object 
 5   Artist              3 non-null      object 
 6   Unit_Price_dollars  3 non-null      float64
 7   Total_runtime_min   3 non-null      float64
 8   song_size_Mb        3 non-null      float64
dtypes: float64(3), int64(2), object(4)
memory usage: 240.0+ bytes


In [95]:
# new runtime mean
from siuba.siu import call

df >> call(np.mean, _.Total_runtime_min)

#np.mean(df["Total_runtime_min"])



3.7936944444444443

In [96]:
#new mean for songsize

df >> call(np.mean, _.song_size_Mb)

#np.mean(df['song_size_Mb'])

7.166306495666504

#### New averages for runtime and song size

The new mean for the runtime is 4.34 minutes reflecting the average time for a song (Harris,2021). 

The new mean for the song size is 7.8 megabytes which also reflects the average size of a song (Harris,2021). 

REFERENCE
Harris, M. (January 14,2021). How Many Songs Does a Gigabyte of Storage Hold? Lifewire: Tech for Humans. <https://www.lifewire.com/how-many-songs-does-a-gigabyte-of-storage-hold-2438442#:~:text=Remember%20to%20multiply%203.5%20by%2060%20to%20get%20the%20number%20of%20seconds.&text=The%20result%20is%20a%20rough,megabytes(MB)%20per%20song.>

In [97]:
# remove bytes and milliseconds column

df = df >> select (~_.Milliseconds, ~_.Bytes)

df.head()

Composer  Genre  \
Song_Track                                                               
Beira Mar                                          Gilberto Gil  Latin   
O Último Romântico (Ao Vivo)                            Unknown  Latin   
Brasil                        Milton Nascimento, Fernando Brant  Latin   

                                                                        Album  \
Song_Track                                                                      
Beira Mar                                                           Unplugged   
O Último Romântico (Ao Vivo)  Lulu Santos - RCA 100 Anos De Música - Álbum 02   
Brasil                                              Milton Nascimento Ao Vivo   

                                         Artist  Unit_Price_dollars  \
Song_Track                                                            
Beira Mar                          Eric Clapton                0.99   
O Último Romântico (Ao Vivo)        Lulu Santos                0.99   
Brasil                        Milton Nascimento                0.99   

                              Total_runtime_min  song_size_Mb  
Song_Track                                                     
Beira Mar                              4.924067      9.153360  
O Último Romântico (Ao Vivo)           3.866550      7.336328  
Brasil                                 2.590467      5.009232

#### Conclusion

This is a complete analysis of itunes data at a given time. A few changes were made to the data given:

1. Checked for missing data and other information about the columns
2. Filled in missing data in Composer column
3. Removed Index and set index to songtrack (The song name)
4. Changed the column name to reflect currency in dollars
5. Changed milliseconds to minutes and sorted in descending order
6. Changed bytes to megabytes
7. Calculated mean for price, runtime and song size
8. Found potential outliers and filtered them out
9. Calculated new mean and made conclusions
10. Removed milliseconds and bytes columns

This data could be visualized using ggplots, matplotlib, seaborn or tableau to visualize people's taste in music at the given time of this data.

In [98]:
# save as csv file

df.to_csv("New_Itunes_data.csv")